A number chain is created by continuously adding the square of the digits in a number to form a new number until it has been seen before.

For example,

44 → 32 → 13 → 10 → 1 → 1  
85 → 89 → 145 → 42 → 20 → 4 → 16 → 37 → 58 → 89

Therefore any chain that arrives at 1 or 89 will become stuck in an endless loop. What is most amazing is that EVERY starting number will eventually arrive at 1 or 89.

How many starting numbers below ten million will arrive at 89?

In [1]:
import sys
sys.path.append('..')
from euler import Progress

## Intuitive implementation
Slow so work on a quicker one below.

In [2]:
squared = lambda x: x*x

def chain_number(num):
    while True:
        if num == 1:
            return False
        if num == 89:
            return True
        num = sum(map(squared, map(int, str(num))))

In [3]:
eighty_nine = 0
bar = Progress()
max_num = 10**7

for i in range(1, max_num):
    if i % 10000 == 0:
        bar.tick((i+1)/max_num)
    eighty_nine += chain_number(i)
    
bar.tick(1)
print(eighty_nine)

Progress: [##############################] 100.0% 116.0 seconds
8581146


## Smarter
If the chain will repeat itself, if we arrive at a previous start number, return that number instead.

In [4]:
squared = lambda x: x*x

memory = {}
def chain_number_memory(num):
    start = num
    while True:
        if num in memory:
            return memory[num]
        if num == 1:
            memory[start] = False
            return False
        if num == 89:
            memory[start] = True
            return True
        num = sum(map(squared, map(int, str(num))))

In [5]:
eighty_nine = 0
bar = Progress()
max_num = 10**7

for i in range(1, max_num):
    if i % 10000 == 0:
        bar.tick((i+1)/max_num)
    eighty_nine += chain_number_memory(i)
    
bar.tick(1)
print(eighty_nine)

Progress: [##############################] 100.0% 75.5 seconds
8581146


### Conclusion
Default method takes 116 s, quicker takes 75 s.  
Decrease with 35% computationt time.